### ISSUE TITLE
"v" is redeclared when mangle is enabled

### ISSUE THREAD

**Issue by Christilut** [terser/terser#927]
**Bug report or Feature request?**

Bug

**Version (complete output of `terser -V` or specific git commit)**

Electron-forge 6.0.0
    "terser-webpack-plugin": "^4.2.3",

**`terser` input**

I'm building an Electron app with Electron-forge and getting the below error.

This happens when I run the `make` command.

The error disappears when I set the `mangle: false` option.
I tried the sub options for mangle, but none seem to make a difference.

**`terser` output or error**
```
√ Checking your system
√ Resolving Forge Config
We need to package your application before we can make it
- Preparing to Package Application for arch: x64You have set packagerConfig.ignore, the Electron Forge webpack plugin normally sets this automatically.

Your packaged app may be larger than expected if you dont ignore everything other than the '.webpack' folder
√ Compiling Main Process Code
× Compiling Renderer Template

An unhandled error has occurred inside Forge:
Compilation errors in the renderer: Hash: 7b5f7a52951a8de03307
Version: webpack 4.44.2
Time: 39923ms
Built at: 2021-02-03 10:19:12
                                 Asset      Size  Chunks               Chunk Names
                     0/index.worker.js   199 KiB
  018aa50ec0c8a8e8d1ac6d238605c676.svg  3.42 KiB          [immutable]
                     1/index.worker.js  62.4 KiB
  1441665c7d54f1c49d1efadce5dc3b0a.svg  1.42 KiB          [immutable]
  1f3523ddeb8414ec1841e1900527e2ce.svg  1.13 KiB          [immutable]
                     2/index.worker.js  64.3 KiB
23ac55ea6fa0fdb4518d9d71cf82a207.woff2   2.8 KiB          [immutable]
                     3/index.worker.js  90.4 KiB
                     4/index.worker.js  54.3 KiB
  4b594e50fe808a7878ca55bf0021c248.png  9.76 KiB          [immutable]
                     5/index.worker.js  89.8 KiB
  54904ce89796c2db44cbea3836ee51b3.gif   110 KiB          [immutable]
                     6/index.worker.js  91.3 KiB
  61a08a1004745f961f1266d87d1186d8.svg  4.69 KiB          [immutable]
 645243e53954237de5248b4ab390ecf7.woff  3.49 KiB          [immutable]
                     7/index.worker.js  91.8 KiB
  942638e66612b31fad1b924a405a29ba.ttf   302 KiB          [immutable]
  a7679a7b608487334b9aa354320c13bc.ttf   225 KiB          [immutable]
  b4cea91c716f97abcc7293b1f412f93f.svg    32 KiB          [immutable]
  caa6bc05f84edd738bdbf256fe19f330.png  8.04 KiB          [immutable]
e9db4005489e24809b62e61177c543a8.woff2   318 KiB          [immutable]
  f91d667cb0849857b962e0e312464017.svg  3.22 KiB          [immutable]
  fe3c77ef606bb90a3b1211bff1dd0680.svg  1.03 KiB          [immutable]
                  main_window/index.js    25 MiB       0               main_window
Entrypoint main_window = main_window/index.js
  [0] ./src/layout/ids.ts 14.4 KiB {0} [built]
  [4] ./src/helpers/log.ts 7.15 KiB {0} [built]
 [15] ./src/helpers/path.ts 7.91 KiB {0} [built]
 [20] ./src/system/sentry.ts + 46 modules 208 KiB {0} [built]
      | ./src/system/sentry.ts 1.77 KiB [built]
      | ./node_modules/@sentry/electron/esm/renderer/index.js 2.78 KiB [built]
      | ./node_modules/@sentry/core/esm/index.js 698 bytes [built]
      | ./node_modules/@sentry/minimal/esm/index.js 6.54 KiB [built]
      | ./node_modules/@sentry/electron/esm/renderer/backend.js 5.02 KiB [built]
      | ./node_modules/@sentry/electron/esm/renderer/client.js 2.08 KiB [built]
      | ./node_modules/@sentry/types/esm/session.js 384 bytes [built]
      | ./node_modules/@sentry/core/esm/integrations/index.js 139 bytes [built]
      | ./node_modules/@sentry/core/esm/sdk.js 574 bytes [built]
      | ./node_modules/@sentry/core/esm/baseclient.js 18.3 KiB [built]
      | ./node_modules/@sentry/types/esm/status.js 1.44 KiB [built]
      | ./node_modules/@sentry/core/esm/transports/noop.js 691 bytes [built]
      | ./node_modules/@sentry/core/esm/basebackend.js 2.08 KiB [built]
      | ./node_modules/@sentry/types/esm/severity.js 1.4 KiB [built]
      | ./node_modules/@sentry/core/esm/request.js 1.95 KiB [built]
      |     + 32 hidden modules
 [24] ./src/classes/Playlist/Playlist.ts 14.8 KiB {0} [built]
 [28] external "electron" 42 bytes {0} [built]
 [40] ./src/classes/HotkeyManager/HotkeyManager.ts + 3 modules 14 KiB {0} [built]
      | ./src/classes/HotkeyManager/HotkeyManager.ts 7.08 KiB [built]
      | ./src/helpers/key.ts 4.67 KiB [built]
      | ./src/classes/HotkeyManager/global/exceptions.ts 937 bytes [built]
      | ./src/classes/HotkeyManager/global/hardcoded.ts 1.33 KiB [built]
 [59] ./src/layout/navigation/playlists/index.ts + 9 modules 45.3 KiB {0} [built]
      | ./src/layout/navigation/playlists/index.ts 26.3 KiB [built]
      | ./src/features/share/index.ts 571 bytes [built]
      | ./src/layout/navigation/playlists/actions/create.ts 2.22 KiB [built]
      | ./src/layout/navigation/playlists/actions/crossreference.ts 2.95 KiB [built]
      | ./src/layout/navigation/playlists/actions/delete.ts 634 bytes [built]
      | ./src/layout/navigation/playlists/actions/duplicate.ts 1.07 KiB [built]
      | ./src/layout/navigation/playlists/actions/merge.ts 868 bytes [built]
      | ./src/layout/navigation/playlists/actions/prefix.ts 6.38 KiB [built]
      | ./src/layout/navigation/playlists/actions/sort.ts 2.82 KiB [built]
      | ./src/layout/navigation/playlists/actions/_helpers.ts 1.44 KiB [built]
 [68] ./src/layout/content/tracks/render.ts 960 bytes {0} [built]
 [77] ./src/classes/HotkeyManager/focus/index.ts 1.18 KiB {0} [built]
[138] ./src/classes/Library/Library.ts 2.39 KiB {0} [built]
[166] ./src/classes/ApplicationWindow/ApplicationWindow.ts 2.37 KiB {0} [built]
[196] ./src/features/integrations/cratehackers/Cratehackers.ts 921 bytes {0} [built]
[379] multi ./src/renderer.ts 28 bytes {0} [built]
[649] ./src/renderer.ts + 20 modules 59.4 KiB {0} [built]
      | ./src/renderer.ts 2.03 KiB [built]
      | ./src/assets/webix/webix.css 394 bytes [built]
      | ./src/assets/fonts/mdi/scss/materialdesignicons.scss 427 bytes [built]
      | ./src/system/ipc.ts 189 bytes [built]
      | ./src/styles/main.scss 385 bytes [built]
      | ./src/system/axios.ts 2.71 KiB [built]
      | ./src/system/firstlaunch.ts 1.73 KiB [built]
      | ./src/layout/_components/_mixins.ts 1.42 KiB [built]
      | ./src/layout/index.ts 3.08 KiB [built]
      | ./src/layout/_components/mixins/helpMixin.ts 2.33 KiB [built]
      | ./src/layout/_components/mixins/settingsMixin.ts 2.63 KiB [built]
      | ./src/layout/content/index.ts 383 bytes [built]
      | ./src/layout/musicplayer/index.ts 3.25 KiB [built]
      | ./src/layout/musicplayer/init.ts 2.63 KiB [built]
      | ./src/layout/_components/editors/hotkeyEditor.ts 3.11 KiB [built]
      |     + 6 hidden modules
    + 671 hidden modules

ERROR in main_window/index.js from Terser
"v" is redeclared [./node_modules/@marshallofsound/webpack-asset-relocator-loader/dist??ref--1!./node_modules/flexsearch/dist/module/flexsearch.js:1,11045][main_window/index.js:52319,10641]
    at ee (C:\Users\Alpha\workspace\lexicon\node_modules\terser\dist\bundle.min.js:1:19541)
    at On.visit (C:\Users\Alpha\workspace\lexicon\node_modules\terser\dist\bundle.min.js:1:120363)
    at On._visit (C:\Users\Alpha\workspace\lexicon\node_modules\terser\dist\bundle.min.js:1:80860)
    at AST_SymbolLet._walk (C:\Users\Alpha\workspace\lexicon\node_modules\terser\dist\bundle.min.js:1:52292)
    at AST_VarDef.<anonymous> (C:\Users\Alpha\workspace\lexicon\node_modules\terser\dist\bundle.min.js:1:66920)
    at On._visit (C:\Users\Alpha\workspace\lexicon\node_modules\terser\dist\bundle.min.js:1:80910)
    at AST_VarDef._walk (C:\Users\Alpha\workspace\lexicon\node_modules\terser\dist\bundle.min.js:1:66886)
    at AST_Let.<anonymous> (C:\Users\Alpha\workspace\lexicon\node_modules\terser\dist\bundle.min.js:1:66297)
    at On._visit (C:\Users\Alpha\workspace\lexicon\node_modules\terser\dist\bundle.min.js:1:80910)
    at AST_Let._walk (C:\Users\Alpha\workspace\lexicon\node_modules\terser\dist\bundle.min.js:1:66218)
Child HtmlWebpackCompiler:
                          Asset      Size  Chunks  Chunk Names
    __child-HtmlWebpackPlugin_0  4.12 KiB       0  HtmlWebpackPlugin_0
    Entrypoint HtmlWebpackPlugin_0 = __child-HtmlWebpackPlugin_0
    [0] ./node_modules/@electron-forge/plugin-webpack/node_modules/html-webpack-plugin/lib/loader.js!./src/index.html 539 bytes {0} [built]
Child WorkerPluginLoader:
                Asset     Size  Chunks  Chunk Names
    0/index.worker.js  199 KiB       0  0
    Entrypoint 0 = 0/index.worker.js
     [0] ./node_modules/threads/dist/common.js 684 bytes {0} [built]
     [1] ./node_modules/threads/dist/transferable.js 882 bytes {0} [built]
     [2] ./node_modules/threads/worker.js 48 bytes {0} [built]
     [3] ./node_modules/symbol-observable/es/ponyfill.js 333 bytes {0} [built]
     [4] ./node_modules/slugify/slugify.js 7 KiB {0} [built]
     [5] ./node_modules/threads/dist/worker/index.js 8.95 KiB {0} [built]
     [6] ./node_modules/is-observable/index.js 186 bytes {0} [built]
     [7] ./node_modules/symbol-observable/es/index.js 403 bytes {0} [built]
     [8] (webpack)/buildin/harmony-module.js 573 bytes {0} [built]
     [9] ./node_modules/threads/dist/serializers.js 1.54 KiB {0} [built]
    [10] ./node_modules/threads/dist/symbols.js 417 bytes {0} [built]
    [11] ./node_modules/threads/dist/types/messages.js 852 bytes {0} [built]
    [12] ./node_modules/threads/dist/worker/implementation.browser.js 989 bytes {0} [built]
    [13] ./node_modules/ts-loader??ref--2!./src/features/trackmatcher/_thread.ts + 5 modules 56.1 KiB {0} [built]
         | ./node_modules/ts-loader??ref--2!./src/features/trackmatcher/_thread.ts 3.55 KiB [built]
         | ./node_modules/fuse.js/dist/fuse.esm.js 40.2 KiB [built]
         | ./src/helpers/string.ts 1.95 KiB [built]
         | ./src/helpers/track.ts 2.84 KiB [built]
         | ./src/constants.ts 6.18 KiB [built]
         | ./src/enums/generic.ts 1.34 KiB [built]
Child WorkerPluginLoader:
                Asset      Size  Chunks  Chunk Names
    4/index.worker.js  54.3 KiB       0  4
    Entrypoint 4 = 4/index.worker.js
     [0] ./node_modules/threads/dist/common.js 684 bytes {0} [built]
     [1] ./node_modules/threads/dist/transferable.js 882 bytes {0} [built]
     [2] ./node_modules/threads/worker.js 48 bytes {0} [built]
     [3] ./node_modules/symbol-observable/es/ponyfill.js 333 bytes {0} [built]
     [4] ./node_modules/threads/dist/worker/index.js 8.95 KiB {0} [built]
     [5] ./node_modules/is-observable/index.js 186 bytes {0} [built]
     [6] ./node_modules/symbol-observable/es/index.js 403 bytes {0} [built]
     [7] (webpack)/buildin/harmony-module.js 573 bytes {0} [built]
     [8] ./node_modules/threads/dist/serializers.js 1.54 KiB {0} [built]
     [9] ./node_modules/threads/dist/symbols.js 417 bytes {0} [built]
    [10] ./node_modules/threads/dist/types/messages.js 852 bytes {0} [built]
    [11] ./node_modules/threads/dist/worker/implementation.browser.js 989 bytes {0} [built]
    [12] ./node_modules/ts-loader??ref--2!./src/features/utilities/playlistoccurrence/_thread.ts + 1 modules 2.26 KiB {0} [built]
         | ./node_modules/ts-loader??ref--2!./src/features/utilities/playlistoccurrence/_thread.ts 1.31 KiB [built]
         | ./src/helpers/assert.ts 964 bytes [built]
```

Any ideas?

**Comment by fabiosantoscode** [comment]
Hello there!

Here's two things to try:

1. Update Terser -- I've noticed that bundle.min.js is minified due to the stack trace containing mangled names. This is no longer the case, to help with debugging issues.
2. Try to isolate the issue to a smaller part of the file -- you should be able to run Terser yourself and figure out what code was running when Terser did crash. You want the raw chunk that webpack is giving to Terser, not the source files. From there on you can remove parts of the file until you have a minimal reproducible example.

**Comment by Christilut** [comment]
Finally got around to this.
Now on version `5.1.1` but a different error:

```
An unhandled error has occurred inside Forge:
Cannot read property 'javascript' of undefined
TypeError: Cannot read property 'javascript' of undefined
    at C:\Users\chris\workspace\my-app\node_modules\terser-webpack-plugin\dist\index.js:572:38
    at SyncHook.eval (eval at create (C:\Users\chris\workspace\my-app\node_modules\tapable\lib\HookCodeFactory.js:19:10), <anonymous>:89:1)
    at SyncHook.lazyCompileHook (C:\Users\chris\workspace\my-app\node_modules\tapable\lib\Hook.js:154:20)
    at Compiler.newCompilation (C:\Users\chris\workspace\my-app\node_modules\webpack\lib\Compiler.js:631:26)
    at C:\Users\chris\workspace\my-app\node_modules\webpack\lib\Compiler.js:667:29
    at AsyncSeriesHook.eval [as callAsync] (eval at create (C:\Users\chris\workspace\my-app\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:4:1)
    at AsyncSeriesHook.lazyCompileHook (C:\Users\chris\workspace\my-app\node_modules\tapable\lib\Hook.js:154:20)
    at Compiler.compile (C:\Users\chris\workspace\my-app\node_modules\webpack\lib\Compiler.js:662:28)
    at C:\Users\chris\workspace\my-app\node_modules\webpack\lib\Compiler.js:321:11
    at Compiler.readRecords (C:\Users\chris\workspace\my-app\node_modules\webpack\lib\Compiler.js:529:11)
    at C:\Users\chris\workspace\my-app\node_modules\webpack\lib\Compiler.js:318:10
    at AsyncSeriesHook.eval [as callAsync] (eval at create (C:\Users\chris\workspace\my-app\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:13:1)
    at AsyncSeriesHook.lazyCompileHook (C:\Users\chris\workspace\my-app\node_modules\tapable\lib\Hook.js:154:20)
    at C:\Users\chris\workspace\my-app\node_modules\webpack\lib\Compiler.js:315:19
    at AsyncSeriesHook.eval [as callAsync] (eval at create (C:\Users\chris\workspace\my-app\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:13:1)
    at AsyncSeriesHook.lazyCompileHook (C:\Users\chris\workspace\my-app\node_modules\tapable\lib\Hook.js:154:20)
```

config in webpack:
```
  optimization: {
    minimize: isProduction,
    minimizer: [new TerserPlugin({
      terserOptions: {
        compress: false,
        mangle: true
      }
    })]
  }
```

**Comment by Christilut** [comment]
Actually this errors happens with mangle and compress false too. Should I open a new issue?

Downgrading to 4.2.3 for now.

**Comment by fabiosantoscode** [comment]
I suppose this new error could be caused by terser-webpack plugin not being updated. Since terser 5 returns a promise, the code that calls it needs to be updated as well.

**Comment by Christilut** [comment]
I think `Cannot read property 'javascript' of undefined` is due to the terser plugin v5 was used with electron forge that included webpack v4.

The original redeclare error has disappeared, not sure why. Probably had something to do with external packages.

Will close this issue.

**Comment by Christilut** [comment]
The original redeclared error was with `mangle=true` but does not appear during build anymore.

However, the same error now appears during runtime with `compress=true` and `mangle=false`:

`Uncaught SyntaxError: Identifier 'v' has already been declared`

**Comment by Christilut** [comment]
I went through all compress options and this seems to be caused by `compress.join_vars=true`. Setting it to false removes the error.

**Comment by fabiosantoscode** [comment]
Could you try to isolate the input that causes this issue? It's not possible to discern what happened from an error message in the output code.

**Comment by Christilut** [comment]
I tried but there were 50 references to `var v =` in the minified output so I couldn't really make out what scope they were in.
Is there a way to compress but not minify? Or another way to make the compressed output somewhat more readable

**Comment by fabiosantoscode** [comment]
If you add the beautify:true format option (which is on the way out but still works) you can have formatted code. By disabling mangling you can retain the variable names (if this fixes the bug, then it's a mangle bug)

**Comment by Christilut** [comment]
Beautify doesn't seem to do anything here and the docs say it is not implemented anymore.

**Comment by fabiosantoscode** [comment]
I'm on my phone but I just gave it a try on the repl, this seems to work:

```
// edit terser options
{
  module: true,
  compress: {},
  mangle: {},
  format: {beautify:true},
  parse: {},
  rename: {},
}
```

**Comment by fabiosantoscode** [comment]
Don't rely on it though, it's only there until I have the time to go through the output and remove all the .indent() and .space() calls

**Comment by Christilut** [comment]
Hmm it doesn't seem to work here.

If I disable minimize then my code is formatted the same as the source.
When I enable minimize, it compresses it without beautify:

```js
  optimization: {
    minimize: isProduction,
    minimizer: [
      new TerserPlugin({
        terserOptions: {
          format: {
            beautify: true
          },
          module: true,
          compress: {
            join_vars: false // If true, this gives an error during runtime
          }, 
          mangle: {
            keep_classnames: true // If false, Shows "unknown target" error during runtime
          }
        }
      })
    ]
  },
  devtool: 'source-map'
```

**Comment by fabiosantoscode** [comment]
Maybe it's best to run the code through a prettifier like prettier then.

**Comment by Christilut** [comment]
Understood! Will report back when I go down this rabbit hole again.

**Comment by fabiosantoscode** [comment]
Thanks for looking into this :) Terser's been having a ton of issues lately, it's hard to keep up and still have free time.
